Included support for indel error detection and read filtering in CollectSamErrorMetrics

michaelgatzen opened the pull request:
- Created new IndelErrorCalculator and IndelErrorMetric classes
- Adjusted CollectSamErrorMetrics to be able to consider insertions and
deletions, as well as number of inserted and deleted bases
- Added indel_length stratifier
- Using base operation information that have been incorporated into htsjdk
- Tests for indel error detection and filtering

### Description

_Give your PR a **concise** yet **descriptive** title_
_Please explain the changes you made here._
_Explain the **motivation** for making this change. What existing problem does the pull request solve?_
_Mention any issues fixed, addressed or otherwise related to this pull request, including issue numbers or hard links for issues in other repos._
_You can delete these instructions once you have written your PR description._

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [ ] Added or modified tests to cover changes and any new functionality
- [ ] Edited the README / documentation (if applicable)
- [ ] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/BooleanSamErrorReadFilterCriterion.java:
```diff
@@ -0,0 +1,22 @@
+        super(comparator, value);
+    }
+
+    private static final Log log = Log.getInstance(CollectSamErrorMetrics.class);
```
unused?

yfarjoun replied:
(and wrong class )

yfarjoun commented:
Yikes! that's some PR! 

could you split this into 3 separate PRs?

1. add indel error.
2. add indel stratifiers.
3. add filtering ability. 

I think that it will make reviewing the code much easier.

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/IndelErrorMetric.java:
```diff
@@ -0,0 +1,47 @@
+     * The number of insertions. Note: This is not the number of bases that have been inserted.
+     */
+    @MergeByAdding
+    public long NUM_INSERTS = 0;
```
should you also count up the number of bases that were inserted?

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/IndelErrorMetric.java:
```diff
@@ -0,0 +1,47 @@
+    public int INSERTS_Q = 0;
+
+    /**
+     * The number of deletions. Note: This is not the number of bases that have been deleted.
```
should you also count up the number of bases that were deleted?

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/NumericSamErrorReadFilterCriterion.java:
```diff
@@ -0,0 +1,28 @@
+        super(comparator, value);
+    }
+
+    private static final Log log = Log.getInstance(CollectSamErrorMetrics.class);
```
unused

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/SamErrorReadFilterCriterion.java:
```diff
@@ -0,0 +1,45 @@
+     * Method to be called in order to check whether this criterion is satisfied
+     * @param stratus Object to check whether it satisfies the criterion
+     */
+    public abstract void checkCriterion(Object stratus);
```
why does stratus not have a type? what are you going to put there?

michaelgatzen replied:
Since stratifiers can output different types, I wanted to keep it as generic as possible. I changed it now to be more strongly typed.

jonn-smith commented:
@yfarjoun Since these features are built upon each other they'll be somewhat difficult to deconflict at this point.  It's a larger PR, but not completely unwieldy so I think it'll be OK.

yfarjoun commented on src/test/java/picard/sam/SamErrorMetric/CollectSamErrorMetricsTest.java:
```diff
@@ -357,5 +530,335 @@ public void testTooManyDirectives() {
+    }
+
+    @Test(dataProvider = "provideForTestIndelErrors")
+    public void testIndelErrors(final String[] readCigars, final String errorSubscript, final int priorQ, IndelErrorMetric expectedMetric) throws IOException {
```
since all your test-cases use priorQ=30, you could remove that from the signature.

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java:
```diff
@@ -143,6 +172,14 @@
+            "can be one of \"=, !=, <, <=, >, >=\" (depending on the datatype). These criteria are conjunctive, "+
+            "i.e. all criteria must be met for the read to be included in the output. Multiple files can be provided which are disjuncitve to each " +
+            "other (i.e. reads are included if they meet either of the filters).", optional = true)
+    public List<File> FILTERS;
```
Can you have a test example where FILTERS is used? it's always good to see how the author envisions parameters being used...

yfarjoun replied:
what's the argument for needing this new filtering DSL?
Also, what does the name of the filter mean? how does the name of the filter get used? 
finally, it isn't clear from the description what is being filtered...what is suffix?

michaelgatzen replied:
1. Some FILTER files are being parsed in ErrorReadFilterTest.java -> testFilterFileParsing. The FILTER argument isn't used directly, I can write an example test for that
1. The idea of the filtering is to being able to output reads that match certain criteria to being able to look at them manually afterwards which, as far as I understand it, is of particular interest for our long read investigations. Let me know if this feature shouldn't be included in master and we'll keep in a separate branch
1. The filter name determines the name of the output file created which lists reads that match that specific filter's criteria. The suffix corresponds to the stratifier name used for the criterion. I updated the description to reflect that.

yfarjoun replied:
Seems like this should be in a separate tool. 
I would use the ENUM name of the stratification rather than the suffix, in the DSL.

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java:
```diff
@@ -309,6 +370,33 @@ protected int doWork() {
+     * @param samFileHeader Header to use for the output BAM file
+     */
+    private void writeOutputReads(SAMFileHeader samFileHeader) {
+        try {
```
use try-with-resources to make sure that opened resources get closed

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java:
```diff
@@ -309,6 +370,33 @@ protected int doWork() {
+        try {
+            SAMFileWriter writer = new SAMFileWriterFactory()
+                    .setCompressionLevel(2)
+                    .makeBAMWriter(samFileHeader, false, new File(OUTPUT + ".outputreads.bam"));
```
use makeWriter

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java:
```diff
@@ -328,30 +416,205 @@ private void writeMetricsFileForAggregator(final BaseErrorAggregation locusAggre
+        }
+
+        if (!outputReadsById.containsKey(readId)) {
+            outputReadsById.put(readId, read);
```
you're going to store all the reads in memory? this seems a bit much!

michaelgatzen replied:
Quite embarrassing mistake really. Fixed it so that reads are written to file immediately.

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java:
```diff
@@ -309,6 +370,33 @@ protected int doWork() {
+            outputReadsById.values().forEach(writer::addAlignment);
+            writer.close();
+
+            for(SamErrorReadFilter filter : samErrorReadFilters) {
```
what is going on here? what are you writing? where is this behavior described? very confused.

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java:
```diff
@@ -44,12 +43,16 @@
+import static shaded.cloud_nio.com.google.common.collect.Iterables.concat;
+
 /**
  * Program to collect error metrics on bases stratified in various ways.
```
please update the javadoc for this tool to reflect the changes that you are proposing to make to it as this no longer describes all that it does....

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java:
```diff
@@ -328,30 +416,205 @@ private void writeMetricsFileForAggregator(final BaseErrorAggregation locusAggre
+     * @param sequenceDictionary The {@link SAMSequenceDictionary} to use to convert the given {@code locus}.
+     * @return An {@link Interval} representing the given {@code locus} or {@code null} if it cannot be converted.
+     */
+    private static Interval locusToInterval(final Locus locus, final SAMSequenceDictionary sequenceDictionary) {
```
is there really no utility method for this? if so, please find an appropriate utils class and put it there.

michaelgatzen replied:
@jonn-smith I couldn't find a utility function for this, do you know of any or a good place to put it? (he wrote this)

jonn-smith replied:
I wrote it because I couldn't find something that did this.  The closest place to put it seems to be in `IntervalListTools`.

yfarjoun replied:
if you use Locatable instead of Locus, you'll be all set.

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java:
```diff
@@ -328,30 +416,205 @@ private void writeMetricsFileForAggregator(final BaseErrorAggregation locusAggre
+     * @return True, if the record has been seen at the previous locus.
+     */
+    @VisibleForTesting
+    protected boolean hasDeletionBeenProcessed(SamLocusIterator.RecordAndOffset deletionRaO, SamLocusIterator.LocusInfo locusInfo) {
```
methods called "has"- something shouldn't have side-effects...meaning that one should be able to call a "has" function twice and get the same result. but this method has side effects...so it should be called "process"-somthing and return true or false depending on whether it has been already processed.

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/ErrorType.java:
```diff
@@ -13,10 +13,10 @@
-            "For those bases, it calculates the error that can be attributable to pre-sequencing, versus during-sequencing.");
-
+            "For those bases, it calculates the error that can be attributable to pre-sequencing, versus during-sequencing."),
+    INDEL_ERROR(IndelErrorCalculator::new, "Only collects insertion or deletion errors.");
```
no need for "only"....
```suggestion
    INDEL_ERROR(IndelErrorCalculator::new, "Collects insertion and deletion errors at the bases provided.");
```

michaelgatzen commented on src/main/java/picard/sam/SamErrorMetric/NumericSamErrorReadFilterCriterion.java:
```diff
@@ -0,0 +1,28 @@
+package picard.sam.SamErrorMetric;
```
Right, added that. There were also some other files in the SamErrorMetric directory that didn't have licenses (e.g. OverlappingErrorMetric), I added a license to those as well if that's ok

michaelgatzen commented on src/main/java/picard/sam/SamErrorMetric/ReadBaseStratification.java:
```diff
@@ -640,6 +829,30 @@ public String getSuffix() {
+    /**
+     * Stratifies by whether or not the current RecordAndOffset is an insertion
+     */
+    public static final IsInsertionStratifier isInsertionStratifier= new IsInsertionStratifier();
```
You're right, the use of this isn't really apparent when the user is only interested in the metrics. With the filtering in mind though, it would be handy to be able to find all insertions of size 2, for example, which would require a criterion such as `indel_length == 2 && is_insertion == true`. I'll leave this open until we have determined if the filtering should be done in a separate tool or not

michaelgatzen replied:
I have removed the IsInsertion and IsDeletion stratifiers as they are no longer needed

michaelgatzen commented on src/main/java/picard/sam/SamErrorMetric/ReadBaseStratification.java:
```diff
@@ -1014,4 +1233,84 @@ private static int stratifyMappingQuality(final SAMRecord sam) {
+            return null;
+        }
+        if(operation == CollectSamErrorMetrics.BaseOperation.Insertion && cigarElement.getOperator() != CigarOperator.I) {
+            log.warn("Wrong CIGAR operator for the given position. This is an error and should be fixed.");
```
This case should never occur (if it does, it would mean an error in the CIGAR string or an error in the code, i.e. an error by me), however, it is seems like a sensible check to make. I now throw an exception instead of a log warning, let me know if you have a better suggestion

michaelgatzen commented:
I now removed all the read output "filtering" from this PR and actually tried to split this PR up into smaller ones as well.
Before I added all the indel error handling, @jonn-smith made some optimizations to the locus iteration. I prepared a separate branch for him, and he'll open another PR for that.
Therefore, this branch only contains the new indel functionality and the new stratifiers. I tried to separate the stratifiers from the indel functionality and make a separate PR for that as well, but some of the new stratifiers depend on the indel features and it seemed simpler to have it in one PR rather than having interdependencies between two PRs. Let me know if that's ok or if you would suggest otherwise.

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java:
```diff
@@ -345,13 +360,82 @@ private static boolean advanceIteratorAndCheckLocus(final PeekableIterator<Varia
+     * whether or not a deletion has already been processed. Note that calling this method will
+     * have the side effect of signaling that the record is processed at this location.
+     *
+     * @param deletionRaO The RecordAndOffset to be checked
```
```suggestion
     * @param deletionRao The RecordAndOffset to be checked
```

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/IndelErrorCalculator.java:
```diff
@@ -0,0 +1,91 @@
+
+        if (operation == CollectSamErrorMetrics.BaseOperation.Insertion) {
+            nInserts++;
+            CigarElement cigarElement = ReadBaseStratification.getIndelElement(recordAndOffset, operation);
```
can this line be extracted from the if statement?

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/IndelErrorCalculator.java:
```diff
@@ -0,0 +1,91 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2018 The Broad Institute
```
2019 here and elsewhere

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/IndelErrorCalculator.java:
```diff
@@ -0,0 +1,91 @@
+
+/**
+ * A calculator that estimates the error rate of the bases it observes for indels only.
+ * Created by jonn on 6/13/19.
```
no need for "created by" line.

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/IndelErrorMetric.java:
```diff
@@ -0,0 +1,93 @@
+     * The number of insertions. Note: This is not the number of bases that have been inserted.
+     */
+    @MergeByAdding
+    public long NUM_INSERTS = 0;
```
```suggestion
    public long NUM_INSERTIONS = 0;
```

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/IndelErrorMetric.java:
```diff
@@ -0,0 +1,93 @@
+     * The (phred) rate of insertions.
+     */
+    @NoMergingIsDerived
+    public int INSERTS_Q = 0;
```
```suggestion
    public int INSERTIONS_Q = 0;
```

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/BaseErrorCalculator.java:
```diff
@@ -36,8 +38,16 @@
     public void addBase(final SamLocusIterator.RecordAndOffset recordAndOffset, final SamLocusAndReferenceIterator.SAMLocusAndReference locusAndRef) {
-        if (!SequenceUtil.isNoCall(recordAndOffset.getReadBase())) {
-            nBases++;
+        // TODO mgatzen should deletions be counted towards total bases?
```
I personally think that only read-bases should count towards the total bases, which will be used as the denominator

michaelgatzen replied:
I agree - thank you

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/BaseErrorCalculator.java:
```diff
@@ -36,8 +38,16 @@
+                nBases++;
+            }
+        } else if (recordAndOffset.getAlignmentType() == AbstractRecordAndOffset.AlignmentType.Insertion) {
+            CigarElement cigarElement = ReadBaseStratification.getIndelElement(recordAndOffset);
```
final

michaelgatzen replied:
fixed

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java:
```diff
@@ -345,13 +349,81 @@ private static boolean advanceIteratorAndCheckLocus(final PeekableIterator<Varia
+    /**
+     * Map of previously seen deletion records, associated with the locus they have been last seen
+     */
+    private final HashMap<SAMRecord, SamLocusIterator.LocusInfo> previouslySeenDeletions = new HashMap<>();
```
Couldn't a single record be involved in more than one deletion?

yfarjoun commented on src/main/java/picard/sam/SamErrorMetric/ReadBaseStratification.java:
```diff
@@ -476,6 +481,74 @@ public String getSuffix() {
+    /**
+     * Stratifies according to the number of indel bases (from CIGAR string) that the read has.
+     */
+    public static class IndelsInReadStratifier extends RecordStratifier<Integer> {
```
Should this be implemented as a sum of two `CigarOperatorsInReadStratifier` (`I` and `D`)? probably not as fast, but less code repetition?

michaelgatzen replied:
Good idea. I moved the actual counting to a private static method so that we don't need to instantiate two Stratifier objects to get the result. (That probably made it more code overall, but less repetition.)

yfarjoun commented on src/test/java/picard/sam/SamErrorMetric/CollectSamErrorMetricsTest.java:
```diff
@@ -357,5 +528,164 @@ public void testTooManyDirectives() {
+
+    @Test(dataProvider = "provideForTestHasDeletionBeenProcessed")
+    public void testHasDeletionBeenProcessed(final String cigarString, final List<Boolean> expected) {
+        Cigar cigar = TextCigarCodec.decode(cigarString);
```
finals

michaelgatzen replied:
fixed
